flink-queryable-client/src/main/java/de/tub/it4bi/modelserving/qs/ALSPredict.java

```diff
@@ -0,0 +1,104 @@
+import org.apache.flink.api.common.typeutils.base.StringSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+
+import java.io.PrintWriter;
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * Created by zis on 06/05/17.
+ */
+public class ALSPredict {
+
+    public static void main(String[] args) throws Exception {
+
+        if (args.length == 0) {
+            throw new IllegalArgumentException("Missing required job ID argument. "
+                    + "Usage: ./de.tub.it4bi.modelserving.qs.ALSPredict <jobID> [jobManagerHost] [jobManagerPort]");
+        }
+        String jobIdParam = args[0];
+
+        // Configuration
+        final String jobManagerHost = args.length > 1 ? args[1] : "localhost";
+        final int jobManagerPort = args.length > 2 ? Integer.parseInt(args[2]) : 6123;
+
+        System.out.println("Using JobManager " + jobManagerHost + ":" + jobManagerPort);
+
+        final JobID jobId = JobID.fromHexString(jobIdParam);
+        final StringSerializer keySerializer = StringSerializer.INSTANCE;
+        final TypeSerializer<Tuple2<String, String>> valueSerializer =
+                TypeInformation.of(new TypeHint<Tuple2<String, String>>() {
+                }).createSerializer(new ExecutionConfig());
+        final Time queryTimeout = Time.seconds(5);
+
+        try (
+                // This helper is for convenience and not part of Flink
+                QueryClientHelper<String, Tuple2<String, String>> client = new QueryClientHelper<>(
+                        jobManagerHost,
+                        jobManagerPort,
+                        jobId,
+                        keySerializer,
+                        valueSerializer,
+                        queryTimeout)) {
+
+            printUsage();
+            ConsoleReader reader = new ConsoleReader();
+            reader.setPrompt("$ ");
+            PrintWriter out = new PrintWriter(reader.getOutput());
+
+            String line;
+            while ((line = reader.readLine()) != null) {
+                String key = line.toUpperCase().trim();
+                out.printf("[info] Querying the model for <user,item> pair '%s'\n", key);
+                try {
+
+                    String tokens[] = key.split(",");
+                    String userID = tokens[0] + "-U";
+                    String itemID = tokens[1] + "-I";
+
+                    Optional<Tuple2<String, String>> userTuple = client.queryState("ALS_MODEL", userID);
+                    Optional<Tuple2<String, String>> itemTuple = client.queryState("ALS_MODEL", itemID);
+
+                    if (userTuple.isPresent() && itemTuple.isPresent()) {
+
+                        // create user vector
+                        double[] userFactors = Arrays.stream(userTuple.get().f1.split(";"))
+                                .mapToDouble(Double::parseDouble)
+                                .toArray();
+                        RealVector userVector = new ArrayRealVector(userFactors);
```

Comment:
It is possible. I was trying in that way in the beginning. The overhead there is deserialisation.
For ALS Model, since model is not complex (just key values), and not sparse, I decided to go with this approach. For Sparse models it makes sense to have a custom object and serialise it.
What do you think?